feat: warn if the node is not syncing

thiagodeev opened the pull request:
What this PR does:
- implement a new abstraction to interact with the feeder
- poll the feeder every 30 seconds
- log a warning if the node is not synced

thiagodeev committed:
feat: implement feeder synchronization for block fetching

- Added a new Feeder struct to manage fetching the latest block from a specified feeder URL.
- Introduced NewFeeder and NewFeederFromContracts functions to create feeder instances based on URLs or contract addresses.
- Integrated feeder polling into the ProcessBlockHeaders function to ensure the node stays in sync with the latest block.
- Updated constants to include feeder URLs and polling interval.

rodrodros commented on validator/feedersync.go:
```diff
@@ -0,0 +1,77 @@
+	}
+}
+
+// Create a new Feeder instance using the [types.ValidationContracts] type
+// to determine the feeder URL.
+func NewFeederFromContracts(
+	contracts *types.ValidationContracts,
+	logger *junoUtils.ZapLogger,
+) *Feeder {
+	switch contracts.Attest.String() {
+	case constants.SepoliaAttestContractAddress:
+		logger.Debug("creating feeder for sepolia")
+
+		return NewFeeder(constants.SepoliaFeederURL, logger)
+	case constants.MainnetAttestContractAddress:
+		logger.Debug("creating feeder for mainnet")
+
+		return NewFeeder(constants.MainnetFeederURL, logger)
+	default:
+		logger.Errorw("unknown attest contract address, defaulting to mainnet feeder",
+			"address", contracts.Attest.String())
+
+		return NewFeeder(constants.MainnetFeederURL, logger) // Default to mainnet feeder
+	}
+}
```
This doesn't look right. Comparing which network are you on based on some address seems extremely flaky. What if the address changes?

Additionally, why if the address is not right you default to something which follows a random heurisitc I believe. We might as well, stop execution with an error – seems clearer to me. 

Let's refactor this `NewFeederFromContracts` there is plenty information throughout the code of which network we are using. We can use either chain id or semething else

thiagodeev replied:
It's not flaky. The variable name is `SepoliaAttestContractAddress` (or `Mainnet...`), so we know it represents the contract in Sepolia. Also, if Starknet changes the address, the entire tool will stop working, so we could say our entire tool is flaky.

>Additionally, why if the address is not right you default to something which follows a random heurisitc I believe. We might as well, stop execution with an error – seems clearer to me.

Since this is a backup/additional feature, I thought it shouldn't make the tool (the main feature) stop working.
The reason to default to mainnet is to ensure the much as possible the mainnet users (where the real money is) get it working. WDYT?

>Let's refactor this NewFeederFromContracts there is plenty information throughout the code of which network we are using. We can use either chain id or semething else

Got it. In the `ProcessBlockHeaders` function, the only way to know which network we are using is through the contract addresses. If we want to use a different value, we'll need to do a small refactor in the code to pass an additional parameter to this function, only because of this. Is that ok for you?

rodrodros commented on validator/feedersync.go:
```diff
@@ -0,0 +1,77 @@
+	pollingTicker *time.Ticker
+}
+
+// Create a new Feeder instance using the feeder URL.
+func NewFeeder(clientURL string, logger *junoUtils.ZapLogger) *Feeder {
+	return &Feeder{
+		client:            feeder.NewClient(clientURL).WithLogger(logger),
+		pollingTicker:     time.NewTicker(constants.FeederPollingInterval * time.Second),
+		latestBlockNumber: 0,
+	}
+}
```
Why are you returning by reference?

thiagodeev replied:
Sorry. I was thinking about preserving the feeder instance, but I forgot we can return the Feeder struct by value and still preserve the feeder instance as a pointer in the Feeder.client field. I'll change it

rodrodros commented on validator/feedersync.go:
```diff
@@ -0,0 +1,77 @@
+	"github.com/NethermindEth/starknet-staking-v2/validator/types"
+)
+
+type Client = feeder.Client
```
Please remove this type alias.

rodrodros commented on validator/validator.go:
```diff
@@ -187,6 +187,7 @@ func RunBlockHeaderWatcher[S signerP.Signer](
 	}
 }
 
+//nolint:gocyclo // Due to handling the main logic + the check if the node is synced
```
Remove no-lint and refactor

rodrodros commented on validator/validator.go:
```diff
@@ -209,59 +210,88 @@ func ProcessBlockHeaders[Account signerP.Signer](
-			)
-			dispatcher.PrepareAttest <- types.PrepareAttest{
-				BlockHash: attestInfo.TargetBlockHash,
+				logger.Errorw("failed to fetch feeder", "error", err.Error())
```
This error message is very ambiguous. Improve it to: "failed to fetch latest block number from feeder".

rodrodros commented on validator/validator.go:
```diff
@@ -209,59 +210,88 @@ func ProcessBlockHeaders[Account signerP.Signer](
+
+				continue
+			}
+			logger.Debugw("node is synced",
+				"feeder latest block", feeder.LatestBlockNumber())
```
format this correctly

```
logger.Debugw(
   "node is synced",
   "feeder latest block",
   feeder.LatestBlockNumber(),
)
```

rodrodros commented on validator/types/starknet.go:
```diff
@@ -21,8 +21,24 @@ func AddressFromString(addrStr string) Address {
 }
 
+// Convert the address to a string with the "0x" prefix and the length of 66.
 func (a *Address) String() string {
-	return (*felt.Felt)(a).String()
+	length := 66
+	hexStr := (*felt.Felt)(a).String()
+
+	// Check if the hex value is already of the desired length
+	if len(hexStr) >= length {
+		return hexStr
+	}
+
+	// Extract the hex value without the "0x" prefix
+	hexValue := hexStr[2:]
+	// Pad zeros after the "0x" prefix
+	paddedHexValue := fmt.Sprintf("%0*s", length-2, hexValue)
+	// Add back the "0x" prefix to the padded hex value
+	paddedHexStr := "0x" + paddedHexValue
+
+	return paddedHexStr
 }
```
Why do you change the `Address` string code.

Also note that is painfully obvious it was AI generated.

thiagodeev replied:
Not AI.

>Why do you change the Address string code.

Because in the new `NewFeederFromContracts` function ([ref](https://github.com/NethermindEth/starknet-staking-v2/blob/4608d3c51e52386c61066a518fc8999767fa38d4/validator/feedersync.go#L36)), when comparing the address string, it was failing because the returned string was different from the one set in the `constants.go` file (the felt is a number after all, it was trimming leading zeros). This method makes the Address type really return a "valid" address.

rodrodros commented on validator/types/starknet.go:
```diff
@@ -21,8 +21,24 @@ func AddressFromString(addrStr string) Address {
+	length := 66
+	hexStr := (*felt.Felt)(a).String()
+
+	// Check if the hex value is already of the desired length
+	if len(hexStr) >= length {
+		return hexStr
+	}
```
What happens if it is 67 or 68. Where are the unit tests for this `String` method

thiagodeev replied:
Not possible. It would not fit in a felt type.
Sorry, forgot to create one

rodrodros commented on validator/types/starknet.go:
```diff
@@ -21,8 +21,24 @@ func AddressFromString(addrStr string) Address {
+// Convert the address to a string with the "0x" prefix and the length of 66.
 func (a *Address) String() string {
-	return (*felt.Felt)(a).String()
+	length := 66
```
Why is `lenght` a variable and not a constant

thiagodeev replied:
No reason at all (to not use `const` instead of `var`). I can change it.
Buf, if you're talking about having an external `const` variable, I don't think it makes sense since this value is used only inside this method. WDYT?

rodrodros commented on validator/types/starknet.go:
```diff
@@ -21,8 +21,24 @@ func AddressFromString(addrStr string) Address {
+	hexValue := hexStr[2:]
+	// Pad zeros after the "0x" prefix
+	paddedHexValue := fmt.Sprintf("%0*s", length-2, hexValue)
+	// Add back the "0x" prefix to the padded hex value
+	paddedHexStr := "0x" + paddedHexValue
```
Really? Do we need this comment here?

thiagodeev replied:
I confess I may have been a little verbose here 😀. I'll change it

rodrodros reviewed (changes requested):
This PR has very low quality and cannot proceed as it is:
* AI generated code with very low quality, comments that are unnecessarily long and sometimes provide zero value since the code intention is perfectly clear.
* Code style that doesn't align with the rest of the project.
* Bad architecture which is not extensible and it is not foul proof. 
* Made an already complex function even more complex without an attempt to refactor.

thiagodeev commented:
@rodrigo-pino  Thanks for the feedback, but you're wrong on one point:
>- AI generated code with very low quality, comments that are unnecessarily long and sometimes provide zero value since the code intention is perfectly clear.

Maybe you're talking mainly about the `Address.String()` method. It was copied manually by me from the `PadZerosInFelt` function in Starknet.go ([ref](https://github.com/NethermindEth/starknet.go/blob/a53a257e654556a540762e30a7b2a0fe9ad432e1/examples/internal/setup.go#L81)). I agree that sometimes I can be more verbose than needed, but this is something about me, it doesn't mean it is an AI-generated code. The only thing that is AI-generated is the commit message, and I've still revised it.

> - Code style that doesn't align with the rest of the project.
> - Bad architecture which is not extensible and it is not foul proof.

I don't think so, but I'll change it to whatever you think is best.

> - Made an already complex function even more complex without an attempt to refactor.

Sorry for that. I was avoiding more code changes since the last time you said I was changing too many things and this PR should have fewer code changes. I'll do this

rodrodros commented:
> It was copied manually by me from the PadZerosInFelt function in Starknet.go ([ref](https://github.com/NethermindEth/starknet.go/blob/a53a257e654556a540762e30a7b2a0fe9ad432e1/examples/internal/setup.go#L81)). I agree that sometimes I can be more verbose than needed, but this is something about me, it doesn't mean it is an AI-generated code. The only thing that is AI-generated is the commit message, and I've still revised it.

That function has deficiencies at several levels. When copying functions as they are from other repos it is important to reference it in the code.

> I don't think so, but I'll change it to whatever you think is best.

Adding a function that creates a feeder based on some contract address is something that has not future proofing as well as making the code more complicated. Harder to maintain and to read. 

> Sorry for that. I was avoiding more code changes since the last time you said I was changing too many things and this PR should have fewer code changes. I'll do this

No need to be sorry. The goal is to have a well implemented feature. The many lines diff is used as an argument when more than one feature (or a feature that is easily divisible) is implemented in one single PR.

rodrodros commented:
If the refactoring means a huge diff, then the refactoring is done as a separate PR before this PR.
